src/passes/TupleOptimization.cpp

```diff
@@ -0,0 +1,357 @@
+//       those constructs. Atm this pass should only remove things that are
+//       definitely worth lowering.
+//
+
+#include <pass.h>
+#include <support/unique_deferring_queue.h>
+#include <wasm-builder.h>
+#include <wasm.h>
+
+namespace wasm {
+
+struct TupleOptimization : public WalkerPass<PostWalker<TupleOptimization>> {
+  bool isFunctionParallel() override { return true; }
+
+  std::unique_ptr<Pass> create() override {
+    return std::make_unique<TupleOptimization>();
+  }
+
+  // Track the number of uses for each tuple local. We consider a use as a
+  // local.get, a set, or a tee. A tee counts as two uses (since it both sets
+  // and gets, and so we must see that it is both used and uses properly).
+  std::vector<Index> uses;
+
+  // Tracks which tuple local uses are valid, that is, follow the properties
+  // above. If we have more uses than valid uses then we must have an invalid
+  // one, and the local cannot be optimized.
+  std::vector<Index> validUses;
+
+  // When one tuple local copies the value of another, we need to track the
+  // index that was copied, as if the source ends up bad then the target is bad
+  // as well.
+  //
+  // This is a symmetrical map, that is, we consider copies to work both ways:
+  //
+  //    x \in copiedIndexed[y]    <==>    y \in copiedIndexed[x]
+  //
+  std::vector<std::unordered_set<Index>> copiedIndexes;
+
+  void doWalkFunction(Function* func) {
+    // If tuples are not enabled, or there are no tuple locals, then there is no
+    // work to do.
+    if (!getModule()->features.hasMultivalue()) {
+      return;
+    }
+    bool hasTuple = false;
+    for (auto var : func->vars) {
+      if (var.isTuple()) {
+        hasTuple = true;
+        break;
+      }
+    }
+    if (!hasTuple) {
+      return;
+    }
+
+    // Prepare global data structures before we collect info.
+    auto numLocals = func->getNumLocals();
+    uses.resize(numLocals);
+    validUses.resize(numLocals);
+    copiedIndexes.resize(numLocals);
+
+    // Walk the code to collect info.
+    super::doWalkFunction(func);
+
+    // Analyze and optimize.
+    optimize(func);
+  }
+
+  void visitLocalGet(LocalGet* curr) {
+    if (curr->type.isTuple()) {
+      uses[curr->index]++;
+    }
+  }
+
+  void visitLocalSet(LocalSet* curr) {
+    if (getFunction()->getLocalType(curr->index).isTuple()) {
+      // See comment above about tees (we consider their set and get each a
+      // separate use).
+      uses[curr->index] += curr->isTee() ? 2 : 1;
+      auto* value = curr->value;
+
+      // We need the input to the local to be another such local (from a tee, or
+      // a get), or a tuple.make.
+      if (auto* set = value->dynCast<LocalSet>()) {
+        assert(set->isTee());
+        validUses[set->index]++;
+        validUses[curr->index]++;
+        copiedIndexes[set->index].insert(curr->index);
+        copiedIndexes[curr->index].insert(set->index);
```

Comment:
This isn't a set of tuples `(x, y)` that we can store by flipping them. We are given an index and need to find all related indexes to it - that is, we know one of `x, y` and want to know  the other (and it is also a set of others, but that's separate).

Reply:
Right, so if we need to store the edges (0, 1), (1, 2), (0, 3), the current code would have:

```
{
  0: {1, 3},
  1: {0, 2},
  2: {1},
  3: {0},
}
```

I'm suggesting that instead we construct this mapping:

```
{
  0: {1, 3},
  1: {2},
}
```
 
From the comment mentioning that this is a bidirectional mapping, this is what I would have expected.

Reply:
Sorry, I'm missing something. Say I have your mapping, and I get the index "2". I want to find the other indexes I need to mark as bad. How do I do that?

Reply:
You would iterate through the map and for each bad key, you would add all the corresponding values to the work list and for each bad value, you would add the corresponding key to the work list. This may not be simpler overall!

Reply:
Ah, but that's `O(map size)` for each index we realize is bad? It's `O(num related indexes)` in the code atm. So I worry changing this would be a regression, though it would save some memory otoh. For now though I think this is good enough.

Reply:
sgtm